core/services/run_queue.go

```diff
@@ -34,81 +36,184 @@ type RunQueue interface {
-		return
-	}
-	rq.workers[runID] = 1
-	numberRunQueueWorkers.Set(float64(len(rq.workers)))
-	rq.workersMutex.Unlock()
+func (rq *runQueue) orchestrateWorkers() {
+	defer close(rq.chDone)
+	for {
+		select {
+		case run := <-rq.chRuns:
+			worker, exists := rq.workers[*run.JobSpecID]
+			if !exists {
+				worker = &singleJobSpecWorker{make(chan *models.JobRun), make(chan struct{}), make(chan struct{})}
+				rq.workers[*run.JobSpecID] = worker
+				atomic.AddInt32(&rq.numWorkers, 1)
+				go rq.runSingleJobSpecWorker(*run.JobSpecID, worker)
+				worker.chAddRun <- run
+			} else {
+				select {
+				case worker.chAddRun <- run:
+				case <-worker.chDone:
+					// If the worker is just spinning down as new work is coming in, allow it
+					// to die and spin up a new one.
+					worker = &singleJobSpecWorker{make(chan *models.JobRun), make(chan struct{}), make(chan struct{})}
+					rq.workers[*run.JobSpecID] = worker
+					atomic.AddInt32(&rq.numWorkers, 1)
+					go rq.runSingleJobSpecWorker(*run.JobSpecID, worker)
+					worker.chAddRun <- run
+				}
+			}
 
-	rq.workersWg.Add(1)
-	go func() {
-		for {
-			rq.workersMutex.Lock()
-			queueCount := rq.workers[runID]
-			if queueCount <= 0 {
-				delete(rq.workers, runID)
-				numberRunQueueWorkers.Set(float64(len(rq.workers)))
-				rq.workersMutex.Unlock()
-				break
+		case jobSpecID := <-rq.chWorkerDone:
+			delete(rq.workers, jobSpecID)
+			if n := atomic.AddInt32(&rq.numWorkers, -1); n < 0 {
+				panic("numWorkers should never be < 0")
 			}
-			rq.workers[runID] = queueCount - 1
-			rq.workersMutex.Unlock()
 
-			if err := rq.runExecutor.Execute(run.ID); err != nil {
-				logger.Errorw(fmt.Sprint("Error executing run ", runID), "error", err)
+		case <-rq.chStop:
+			for _, worker := range rq.workers {
+				worker.chStop <- struct{}{}
+				<-worker.chDone
+				if n := atomic.AddInt32(&rq.numWorkers, -1); n < 0 {
+					panic("numWorkers should never be < 0")
+				}
 			}
+			return
 		}
+	}
+}
+
+func (rq *runQueue) runSingleJobSpecWorker(jobSpecID models.ID, worker *singleJobSpecWorker) {
+	defer close(worker.chDone)
+
+	promNumberRunQueueWorkers.Inc()
+
+	var (
+		startOnce       sync.Once
+		chWorkerStarted = make(chan struct{})
+		chResume        = make(chan []models.ID)
+		chWorkComplete  = make(chan struct{})
+	)
+
+	// The worker goroutine accepts job run requests from the coordinator loop below.  It
+	// can accept further requests after work has begun.  Once its queue of work is empty,
+	// it sends a "work finished" message to that loop, which may or may not cause it to
+	// spin down, depending on whether further requests are already enqueued.
+	go func() {
+		for runsRequested := range chResume {
+			startOnce.Do(func() { close(chWorkerStarted) })
+
+			for _, runID := range runsRequested {
+				select {
+				case <-worker.chStop:
+					return
+				default:
+				}
+
+				promNumberRunsQueued.Inc()
 
-		rq.workersWg.Done()
+				if err := rq.runExecutor.Execute(&runID); err != nil {
```

Comment:
I'm not certain this achieves the requirement of serializing the complete execution of the runs for a given job, because the `RunExecutor` can [leave a run in a pending state](https://github.com/smartcontractkit/chainlink/blob/fix-racy-tests/core/services/run_executor.go#L72), if there aren't yet enough block confirmations, for instance.